### ISSUE TITLE
NXP LS Platforms: pkcs_1014 test is failing

### ISSUE THREAD

**Issue by sahilnxp** [OP-TEE/optee_os#4909]
Hi,

I am working on latest OP-TEE code with commit https://github.com/OP-TEE/optee_os/commit/2de2880065f3f8232873dfb1dd4b94322f0f3971

During testing pkcs_1014 is failing with following error:
```
* pkcs11_1014 PKCS11: Test C_SetAttributeValue()
o pkcs11_1014.1 Set attributes on secret key object
/home/nxa19713/linux-factory/optee-test/host/xtest/pkcs11_1000.c:3496: Expression "g_decrypt == CK_TRUE" (0 == 1) is false
  pkcs11_1014.1 FAILED
  pkcs11_1014 FAILED
```

When i tested with 3.13.0 tag it is working fine.

I checked the code there are no changes related to functions being called during this test in optee-os, optee-client and optee-test after 3.13.0 tag also I don't think there is any platform specific thing being done during this test.

@etienne-lms @vesajaaskelainen @ruchi393 Please have a look at this and let me know if I am missing something here.

Regards,
Sahil Malhotra

**Comment by jforissier** [comment]
Hello Sahil,

I'm answering not because I know about the PKCS#11 stuff, but because I troubleshooted a similar issue recently (https://github.com/OP-TEE/optee_test/issues/551). I would recommend the following steps. First checkout optee_os, optee_client, optee_test 3.13.0 and test again to make sure it works as you said. Then checkout only optee_os master and test. If it fails, bisect optee_os, otherwise checkout optee_client master and test. If it fails, bisect optee_client, otherwise checkout optee_test master and test again, it should fail, then git bisect.

**Comment by WindowsXp-Beta** [comment]
I encountered the same problem when testing pkcs11_1014 on hikey960.
```
* pkcs11_1014 PKCS11: Test C_SetAttributeValue()
o pkcs11_1014.1 Set attributes on secret key object
external/optee_test/host/xtest/pkcs11_1000.c:3278: Expression "g_unwrap == CK_TRUE" (0 == 1) is false
  pkcs11_1014.1 FAILED
  pkcs11_1014 FAILED
```
I used a pre-built from [here](https://people.linaro.org/~victor.chong/prebuilt/pie/3130/hikey960/) and only this test failed.
<img width="599" alt="BF90C212-C4A1-4483-8DC6-50DD0880DB5B" src="https://user-images.githubusercontent.com/65265345/136680080-bb43f1f1-3458-4208-974f-1648272e1521.png">

**Comment by vesajaaskelainen** [comment]
`external/optee_test/host/xtest/pkcs11_1000.c:3278` -- there is no code at that line. And that line would be in 1013 examples -- please update your repos. @vchong seems to be your (?) prebuilt stuff perhaps you have advice for that case.

**Comment by vesajaaskelainen** [comment]
@sahilnxp Synchronized my test tree with latest masters:
```
+-----------------------------------------------------
3640 subtests of which 0 failed
25 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
```
This was on Xilinx MPSoC (Cortex-A53 64bit) QEMU.

Test code around there is only doing create secret object (line 3480) in session (eg. does not go to secure storage) and it more or less verifies that was it created correctly (from line 3487). Next step would have been to actually modify it.

Are you sure you have updated optee-os and optee-client also to match?

Otherwise it starts to sound that memory has been corrupted somewhere.

You may want to enable core and ta traces and see if those gives any more clues. (I had those enabled in my test bench)

**Comment by sahilnxp** [comment]
Thanks everyone for looking at this issue, looks like there is some issue with optee-test.

We had a set of images based on 3.13.0 tag which are compiled using Yocto and with these images, pkcs11_1014 was passing successfully, that's why I said the 3.13.0 release it was working fine.
But when I compiled 3.13.0 standalone, pkcs11_1014  was failing.
Both yocto and standalone compilation was using gcc 10.2 toolchain.
So from this it looks like there is some issue in compilation environment between yocto and standalone compilation.

So I dig into the yocto recipes, in those recipes we were using -O2 flag during optee-test compilation.
I used the same CFLAGS in standalone compilation and tested with that xtest, in that case pkcs11_1014 test passed.

So. from this we can say that -O2 CFLAG is doing some optimizations in code, which leads to pkcs11_1014 to pass.
